src/explore-education-statistics-common/src/components/ContentSectionIndex.tsx

```diff
@@ -0,0 +1,89 @@
+import React, { Component } from 'react';
+import findAllForId from '@common/lib/dom/findAllForId';
+
+interface IndexResult {
+  element: Element;
+  location: string;
+}
+
+interface Props {
+  fromId: string;
+}
+
+interface State {
+  indexResults: IndexResult[];
+  indexComplete: boolean;
+}
+
+class ContentSectionIndex extends Component<Props> {
+  public state: State = {
+    indexResults: [],
+    indexComplete: false,
+  };
+
+  public componentDidMount() {
+    const { fromId } = this.props;
+    this.index(fromId);
+  }
+
+  private static getLocationText(element: HTMLElement): string {
+    const location: string[] = [];
+
+    element.tagName.toLowerCase();
+    location.unshift(element.textContent || '');
+    return location.join(' > ');
+  }
+
+  private index = (fromId: string) => {
+    const elements = findAllForId(fromId, 'h2,h3,h4');
+
+    const indexResults =
+      elements != null
+        ? elements.map(element => {
+            const location = ContentSectionIndex.getLocationText(element);
+
+            return {
+              element,
+              location,
+            };
+          })
+        : [];
+
+    this.setState({ indexResults, indexComplete: true });
+  };
+
+  public render() {
+    const { indexResults, indexComplete } = this.state;
+
+    return (
+      <>
+        {indexComplete && <h3 className="govuk-heading-s">In this section</h3>}
+
+        <ul className="govuk-body-s">
+          {indexComplete &&
+            indexResults.map((item, index) => {
+              return (
+                <>
+                  {item.location && (
+                    <li>
+                      <a
+                        href="#"
+                        onClick={e => {
```

Comment:
This will not be a sufficient solution in non-Javascript land.

We need the links to have standard hrefs that reference the specific element ID or names on the page e.g. `<a href="#my-element">`.

If the element doesn't have an ID, we can generate a kebab-cased one from that element's text content (use the Lodash `kebabCase`) and attach it manually via `setAttribute`.

Reply:
I think we'll probably be able to change this once the content editor is done & we can assign ids to things but for now this is just imported html content.

Reply:
Looking at some of the imported HTML content, it looks like a bunch of the headings already have IDs attached to them, so why not assign these into the hrefs?

Reply:
It's also worth noting that it doesn't look like this current implementation will work in non-JS at all as we really need to ideally receive the heading IDs from the API response.

Reply:
K - good spot - have added id in href.